bitacross-worker/bitacross/core/bc-task-receiver/src/lib.rs

```diff
@@ -123,51 +146,59 @@ impl<
 		RRL: RelayerRegistryLookup,
 		ERL: EnclaveRegistryLookup,
 		SRL: SignerRegistryLookup,
-	> BitAcrossTaskContext<SKR, EKR, BKR, S, H, O, RRL, ERL, SRL>
+		Responder,
+	> BitAcrossTaskContext<SKR, SIGNINGAK, EKR, BKR, S, H, O, RRL, ERL, SRL, Responder>
 where
-	SKR: AccessKey,
+	SKR: AccessKey + AccessPubkey<KeyType = Rsa3072PubKey>,
+	SIGNINGAK: AccessKey<KeyType = ed25519::Pair>,
 	EKR: AccessKey<KeyType = EcdsaPair>,
 	BKR: AccessKey<KeyType = SchnorrPair>,
 	<SKR as AccessKey>::KeyType: ShieldingCryptoEncrypt + 'static,
 	H::StateT: SgxExternalitiesTrait,
+	Responder: SendRpcResponse<Hash = H256>,
 {
 	#[allow(clippy::too_many_arguments)]
 	pub fn new(
 		shielding_key: Arc<SKR>,
+		signing_key_access: Arc<SIGNINGAK>,
 		ethereum_key_repository: Arc<EKR>,
 		bitcoin_key_repository: Arc<BKR>,
 		enclave_signer: Arc<S>,
 		state_handler: Arc<H>,
 		ocall_api: Arc<O>,
 		relayer_registry_lookup: Arc<RRL>,
-		musig2_ceremony_registry: Arc<Mutex<CeremonyRegistry<BKR>>>,
 		enclave_registry_lookup: Arc<ERL>,
 		signer_registry_lookup: Arc<SRL>,
-		musig2_ceremony_pending_commands: Arc<Mutex<CeremonyCommandsRegistry>>,
 		signing_key_pub: [u8; 32],
+		ceremony_registry: Arc<RwLock<CeremonyRegistry<BKR>>>,
+		ceremony_command_tmp: Arc<RwLock<CeremonyCommandTmp>>,
+		responder: Arc<Responder>,
 	) -> Self {
 		Self {
 			shielding_key,
+			signing_key_access,
 			ethereum_key_repository,
 			bitcoin_key_repository,
 			enclave_signer,
 			state_handler,
 			ocall_api,
 			relayer_registry_lookup,
-			musig2_ceremony_registry,
 			enclave_registry_lookup,
 			signer_registry_lookup,
-			musig2_ceremony_pending_commands,
 			signing_key_pub,
+			ceremony_registry,
+			ceremony_command_tmp,
+			responder,
 		}
 	}
 }
 
 #[allow(clippy::type_complexity)]
-pub fn run_bit_across_handler_runner<SKR, EKR, BKR, S, H, O, RRL, ERL, SRL>(
-	context: Arc<BitAcrossTaskContext<SKR, EKR, BKR, S, H, O, RRL, ERL, SRL>>,
+pub fn run_bit_across_handler_runner<SKR, SIGNINGAK, EKR, BKR, S, H, O, RRL, ERL, SRL, Responder>(
```

Comment:
Previously this function was responsible just for general task handling which was fine in my opinion.
Now it's getting coupled to musig2 ceremonies - can we avoid that ?

Previously:

* `bitacross-worker/litentry/core/direct-call/src/handler/sign_bitcoin.rs`
* `bitacross-worker/litentry/core/direct-call/src/handler/nonce_share.rs`
* `bitacross-worker/litentry/core/direct-call/src/handler/partial_signature_share.rs`
* `bitacross-worker/litentry/core/direct-call/src/handler/kill_ceremony.rs`

Where the only units from direct-call processing aware of musig ceremonies

Reply:
I would prefer to keep them separated - not they are mixed - for example it will be hard to extend in the future. 
I assume we sacrifice readability for performance ?

Renaming things won't change much in this case.

Reply:
Could you encapsulate added logic in `handle_ceremony_command function` ?

Reply:
OK